packages/blade/src/components/Menu/useMenu.ts

```diff
@@ -0,0 +1,183 @@
+  useFloatingParentNodeId,
+  useFloatingTree,
+  useHover,
+  useInteractions,
+  useListNavigation,
+  useListItem,
+  useRole,
+  useTransitionStyles,
+} from '@floating-ui/react';
+import type { MenuContextType, UseFloatingMenuProps, UseFloatingMenuReturnType } from './types';
+import { useControllableState } from '~utils/useControllable';
+import { OVERLAY_OFFSET, OVERLAY_TRANSITION_OFFSET } from '~components/BaseMenu/tokens';
+import { makeSize, useTheme } from '~utils';
+
+const MenuContext = React.createContext<MenuContextType>({
+  getItemProps: () => ({}),
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  setHasFocusInside: () => {},
+  isOpen: false,
+});
+
+const useMenu = (): MenuContextType => {
+  const contextValue = React.useContext(MenuContext);
+  return contextValue;
+};
+
+const useFloatingMenuSetup = ({
+  elementsRef,
+  openInteraction,
+  onOpenChange,
+  isOpen,
+}: UseFloatingMenuProps): UseFloatingMenuReturnType => {
+  const [isControllableOpen, setIsControllableOpen] = useControllableState({
+    value: isOpen,
+    defaultValue: false,
+    onChange: (isOpen) => onOpenChange?.({ isOpen }),
+  });
+  const [activeIndex, setActiveIndex] = React.useState<number | null>(null);
+
+  const tree = useFloatingTree();
+  const nodeId = useFloatingNodeId();
+  const parentId = useFloatingParentNodeId();
+  const item = useListItem();
+  const { theme } = useTheme();
+
+  const isNested = parentId != null;
+
+  const { floatingStyles, refs, context } = useFloating<HTMLButtonElement>({
+    nodeId,
+    open: isControllableOpen,
+    onOpenChange: (_isOpen) => setIsControllableOpen(() => _isOpen),
+    placement: isNested ? 'right-start' : 'bottom-start',
+    middleware: [
+      offset({ mainAxis: isNested ? 12 : OVERLAY_OFFSET, alignmentAxis: isNested ? -16 : 0 }),
+      flip(),
+      shift(),
+    ],
+    whileElementsMounted: autoUpdate,
+  });
+
+  const hover = useHover(context, {
+    enabled: isNested || openInteraction === 'hover',
+    delay: { open: 75 },
+    handleClose: safePolygon({ blockPointerEvents: true }),
+  });
+  const click = useClick(context, {
+    event: 'mousedown',
+    toggle: !isNested,
+    ignoreMouse: isNested,
+  });
+  const role = useRole(context, { role: 'menu' });
+  const dismiss = useDismiss(context, { bubbles: true });
+  const listNavigation = useListNavigation(context, {
+    listRef: elementsRef,
+    activeIndex,
+    nested: isNested,
+    onNavigate: setActiveIndex,
+    focusItemOnHover: false,
+  });
+
+  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions([
+    hover,
+    click,
+    role,
+    dismiss,
+    listNavigation,
+  ]);
+
+  const { isMounted, styles: floatingTransitionStyles } = useTransitionStyles(context, {
+    duration: theme.motion.duration.quick,
+    initial: ({ side }) => {
+      let transform: string;
+      const transitionOffset = makeSize(OVERLAY_TRANSITION_OFFSET);
+      switch (side) {
```

Comment:
nit: use ifelse or object mapper. 

Switch case is making the code bit longer than it needs to be.

Reply:
makes sense. done in upcoming commit